# Ej 47

La mediana de cantidad de links internos que tienen **todos los contenidos que existen**

## Supuestos

Suponiendo que **todos los contenidos que existen en Wikipedia se encuentran en `content.csv`**, entonces determino la base de datos que deriva de la interseccion entre `pagelinks_sample.csv` y `content.csv` y sobre la misma determino la mediana de la cantidad de links internos.

Es decir, todos los links internos que aparecen en `pagelinks_sample.csv` van a estar contendio en una pagina cuyo ID se van a encontrar en `content.csv`, pero la nueva informacion que me da son los enlaces internos que tiene la pagina. 

Si una pagina no tiene enlaces internos, entonces el total de enlaces internos sera 0.

In [1]:
# importacion general de librerias y de visualizacion (matplotlib y seaborn)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

plt.style.use('default') # haciendo los graficos un poco mas bonitos en matplotlib
#plt.rcParams['figure.figsize'] = (20, 10)

sns.set(style="whitegrid") # seteando tipo de grid en seaborn

pd.options.display.float_format = '{:20,.2f}'.format # suprimimos la notacion cientifica en los outputs

import warnings
warnings.filterwarnings('ignore')

from datetime import datetime

In [2]:
# Cargo los links internos
pagelinks=pd.read_csv('./Wikipedia_dataset_2c2021/pagelinks_sample.csv',usecols=['pl_from','pl_title'],
                      nrows=10000000)

## 1. Hallo la base de datos con la cantidad de enlaces internos por contenido y muestro el resultado

In [3]:
db_links_internos = pagelinks.groupby('pl_from').agg(links_internos=('pl_title','count'))
display(db_links_internos)

links_internos
pl_from                 
5                    263
7                     45
10                   209
15                    13
17                     4
...                  ...
10010768               1
10010770               3
10010771               1
10010774               4
10010782               1

[3039788 rows x 1 columns]

## 2. Uno la base de datos hallada con el contents.csv

Uno las bases de datos teniendo en cuenta el ID en comun que tiene cada publicacion. 

Como db_pagelinks es un subconjunto de db_contents, entonces hago un inner join.

In [4]:
contents=pd.read_csv('./Wikipedia_dataset_2c2021/contents.csv', usecols=['id']) 
display(contents)

id
0               5
1               7
2              10
3              15
4              17
...           ...
4132159  10010593
4132160  10010595
4132161  10010596
4132162  10010598
4132163  10010600

[4132164 rows x 1 columns]

In [5]:
content_con_links_internos = contents.merge(db_links_internos, how='inner', left_on='id',right_index=True)
display(content_con_links_internos)

id  links_internos
0               5             263
1               7              45
2              10             209
3              15              13
4              17               4
...           ...             ...
4132153  10010587               2
4132157  10010591               2
4132158  10010592               2
4132160  10010595               1
4132163  10010600               3

[1803021 rows x 2 columns]

## Reemplazo los Nans por 0

In [6]:
content_con_links_internos['links_internos']=content_con_links_internos['links_internos'].fillna(0)

## Calculo el valor de la mediana

In [7]:
content_con_links_internos['links_internos'].median()

3.0

In [8]:
content_con_links_internos['links_internos'].size

1803021

# Conclusion

Para una muestra de $1803021$ articulos de wikipedia, se observa que la mediana de cantidad de links internos por articulo, se encuentra en $3.00$.

Es importante destacar que esta muestra de datos pagelinks_sample se encuentra incompleta por motivos limitantes de memoria, por lo que la mediana se podria ser otro valor.